### PR TITLE
refactor(canvas-render): extract side-choice preferences into a named rules-as-data framework

### PR DESCRIPTION
**Pull request by kamiazya**
## What (plan slice S7 — the final rendering-foundation slice; dev-loop implemented, review clean — 0 findings)

Implements decision #10's rule taxonomy: routing-quality feedback now lands as **one named rule + its test** instead of a new branch in an existing function.

- `layout/edge-rules.ts` declares `SIDE_PREFERENCE_RULES` as ordered data — `zeroBendFacingFirst` (lane window ≥20px + facing gap), `dominantAxisFirst`, `lPairCrowdingTieBreak`, `uHookWhenDegenerate`, `gapValidOpposingBeforeInvalid`, `incumbentWinsTies` — each with its own focused unit tests (boundary cases included).
- `rankedSidePairs` becomes a thin composer over the list; the solver (optimizer loop, incremental matrix, sweep build, worst-offender budgeting, locked sets) keeps zero geometric domain knowledge.
- Framework invariants as fast-check properties: candidate generation is **total** (never empty, including dx=dy=0 and interpenetrating rects), **deterministic**, and every preference rule is **order-only** (removing it never adds/removes candidates).
- Behavior-preserving: all ~45 routing/side-choice/parity pins pass **unweakened** per extraction commit; integrator spot-checks — swapping two rules in the list turns the suite red (the pins catch equilibrium changes), and the apps/web browser pins (edge-group-routing, live-drag side tracking, connect preview) plus the pixel goldens pass unchanged.

**Scoped follow-up (timebox clause):** penalty-term extraction (tiered scorers over the shared narrow phase) — filed as `penalty-rules-extraction` on the board; the preference side landed complete rather than both sides rushed.

With this, the rendering-foundation initiative's required scope is fully landed: decision doc (#680), pixel goldens (#687), sweepline + bounded trials (#684), wrap-parity pin (#685), seeded PRNG (#689), and this rule framework.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01TZyjZ1ZozdMU3HQKjvY7pJ

## Summary by CodeRabbit

* **Improvements**
  * Improved edge routing to prioritize cleaner, lower-bend connections.
  * Preserved existing routes when alternatives have equal cost.
  * Improved handling of overlaps, crowded corners, opposing connections, and degenerate paths.

* **Quality**
  * Added comprehensive automated coverage for routing preferences, ordering, determinism, and candidate selection.
  * Documented the routing decision rules and follow-up areas.

### CODE REVIEW

File: packages/canvas-render/src/layout/edge-rules.properties.test.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+// (dx or dy === 0), diagonal (both nonzero), interpenetrating rects
+// (overlapping ranges within [-100,100]/[10,120]), and disjoint rects.
+const contextArb: fc.Arbitrary<PreferenceRuleContext> = fc
+  .record({
+    dx: fc.integer({ min: -300, max: 300 }),
+    dy: fc.integer({ min: -300, max: 300 }),
+    fromRect: rectArb,
+    toRect: rectArb,
+  })
+  .map(({ dx, dy, fromRect, toRect }) => ({ dx, dy, fromRect, toRect, crowd: () => 0 }))
```
[the remark this answers is not in the document]
Comment (kamiazya) [reply packages/canvas-render/src/layout/edge-rules.properties.test.ts:36]: Valid — fixed forward: the generators now floor w/h at 0 (zero-size boxes are legal per canvas-model's own pin), and all 22 rule-framework tests hold on the widened domain. Follow-up PR referenced above.
